### ISSUE TITLE
Bad BGCF config makes sprout cyclically crash

### ISSUE THREAD

**Issue by AlexHockey** [Metaswitch/sprout#1208]
I was editing my BGCF config and I made a typo:

```
{
    "routes" : [
        {
            "name": "Default route",
            "domain": "*",
            "routes": [ "sip:ajh-dev.cw-ngv.com:5080;transport=tcp;lr" ]
        }
    ]
}
```

The third key in the object was "routes" and it is supposed to be "route". When I reloaded sprout it immediately crashed with the following stack. 

```
Thread 1 (Thread 0x7fb520895780 (LWP 17314)):
#0  0x00007fb51d524b99 in __libc_waitpid (pid=17458, stat_loc=stat_loc@entry=0x7fffc1984c20, options=options@entry=0) at ../sysdeps/unix/sysv/linux/waitpid.c:40
#1  0x00007fb51d4aa2e2 in do_system (line=<optimized out>) at ../sysdeps/posix/system.c:148
#2  0x00000000004d2fc3 in Logger::backtrace (this=0x1396be0, data=<optimized out>) at /var/lib/jenkins/workspace/sprout/modules/cpp-common/src/logger.cpp:288
#3  0x000000000058642d in Log::backtrace (fmt=fmt@entry=0x71ff48 "Signal %d caught") at /var/lib/jenkins/workspace/sprout/modules/cpp-common/src/log.cpp:185
#4  0x00000000005e7b6c in signal_handler (sig=6) at main.cpp:1084
#5  <signal handler called>
#6  0x00007fb51d49acc9 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#7  0x00007fb51d49e0d8 in __GI_abort () at abort.c:89
#8  0x00007fb51d493b86 in __assert_fail_base (fmt=0x7fb51d5e4830 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", assertion=assertion@entry=0x6eb375 "false", file=file@entry=0x6eb790 "/var/lib/jenkins/workspace/sprout/modules/rapidjson/include/rapidjson/document.h", line=line@entry=821, function=function@entry=0x6ec220 <rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> >& rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> >::operator[]<rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> >(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> > const&)::__PRETTY_FUNCTION__> "rapidjson::GenericValue<Encoding, Allocator>& rapidjson::GenericValue<Encoding, Allocator>::operator[](const rapidjson::GenericValue<Encoding, SourceAllocator>&) [with SourceAllocator = rapidjson::Mem"...) at assert.c:92
#9  0x00007fb51d493c32 in __GI___assert_fail (assertion=0x6eb375 "false", file=0x6eb790 "/var/lib/jenkins/workspace/sprout/modules/rapidjson/include/rapidjson/document.h", line=821, function=0x6ec220 <rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> >& rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> >::operator[]<rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> >(rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> > const&)::__PRETTY_FUNCTION__> "rapidjson::GenericValue<Encoding, Allocator>& rapidjson::GenericValue<Encoding, Allocator>::operator[](const rapidjson::GenericValue<Encoding, SourceAllocator>&) [with SourceAllocator = rapidjson::Mem"...) at assert.c:101
#10 0x0000000000512d29 in operator[]<rapidjson::MemoryPoolAllocator<> > (name=<synthetic pointer>, this=0x5) at /var/lib/jenkins/workspace/sprout/modules/rapidjson/include/rapidjson/document.h:821
#11 rapidjson::GenericValue<rapidjson::UTF8<char>, rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> >::operator[] (this=this@entry=0x1b1cd84, name=name@entry=0x7097df "route") at /var/lib/jenkins/workspace/sprout/modules/rapidjson/include/rapidjson/document.h:809
#12 0x000000000057cdc9 in operator[] (name=0x7097df "route", this=0x1b1cd84) at /var/lib/jenkins/workspace/sprout/modules/rapidjson/include/rapidjson/document.h:811
#13 BgcfService::update_routes (this=this@entry=0x1b0b530) at bgcfservice.cpp:124
#14 0x000000000057bcaf in operator() (__p=0x1b0b530, this=<synthetic pointer>) at /usr/include/c++/4.8/bits/stl_function.h:551
#15 Updater (run_on_start=true, signal_waiter=<optimized out>, myFunctor=..., pointer=0x1b0b530, this=0x1b166c0) at /var/lib/jenkins/workspace/sprout/modules/cpp-common/include/updater.h:61
```

**Comment by AlexHockey** [comment]
Has probably existed for a while, but marking as critical as if you uploaded this to etcd it would bring the entire cluster down.

**Comment by bossmc** [comment]
Fixed by #1211.
